### PR TITLE
Model generalization

### PR DESCRIPTION
**Pull request by rrevoid**

### CODE REVIEW

File: scripts/train_model.py [hunk 1]
```diff
@@ -14,29 +15,85 @@ def configure_arg_parser():
...
     parser.add_argument("-v", "--val", help="Path to dataset for validation")
     parser.add_argument("-bs", "--batch_size", type=int, default=64, help="Batch size for training")
-    parser.add_argument("-cl", "--max_context", type=int, default=256, help="Max length of context fed into diffusion")
+    parser.add_argument(
+        "-cl",
+        "--max_context",
+        type=int,
+        default=256,
+        help="Max length of context fed into diffusion",
+    )
```
Comment (SpirinEgor) [inline scripts/train_model.py]: Make sure you are using `black` with `-l 120` argument. These statements can be fitted in one line.

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from setuptools import setup, find_packages
+
+setup(
```
Comment (SpirinEgor) [inline setup.py]: Why do you add this?

File: scripts/train_model.py [hunk 3]
```diff
@@ -14,29 +15,85 @@ def configure_arg_parser():
...
-        val_dataset, batch_size=batch_size, collate_fn=lambda x: val_dataset.collate_fn(x, False)
+        val_dataset,
+        batch_size=batch_size,
+        collate_fn=lambda x: val_dataset.collate_fn(x, False),
+    )
+
+    encoder, decoder, emb_dim = get_components(name)
+
+    model = DiDi(encoder, decoder, emb_dim, train_dataset.vocab_size, diffusion_steps, schedule)
+    trainer = pl.Trainer(
```
Comment (SpirinEgor) [inline scripts/train_model.py]: It is very usefull to setup:
- `WandBLogger`
- `ModelCheckpoint` callback
- `val_check_interval` to evaluate model every X steps
- `gradient_clip_val`

And I also prefer to add `LearningRateMonitor("step")` to monitor current learning rate.

File: src/diffusion/model.py [hunk 4]
```diff
@@ -64,3 +86,80 @@ def forward(
...
+            time_ids=ones,
+        )
+
+        pad_mask = gt == self.emb.padding_idx
+
+        probs = self.classifier(x_0)
+
+        loss = calculate_batch_ce(probs, gt, pad_mask).item()
+
+        self.log("val_loss", loss)
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: This way you will log eval loss on each batch. But during evaluation we often want to see average value: how good is the model on average. To do this, you need to return loss and implement `validation_epoch_end` where you will stack all outputs and calculate mean.

Comment (SpirinEgor) [reply src/diffusion/model.py]: And for logging to W&B, I suggest to use grouping. Passing metrics in format `val/loss`, `train/loss` will be splitter to different tabs on frontend.

File: src/diffusion/model.py [hunk 5]
```diff
@@ -64,3 +86,80 @@ def forward(
...
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.AdamW(self.parameters(), lr=0.0001)
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: Pass learning rate to model constructor and use `self.lr` here.

File: src/diffusion/model.py [hunk 6]
```diff
@@ -64,3 +86,80 @@ def forward(
...
+
+        loss = mse + ce + t0_loss
+
+        metrics = {"mse": mse.item(), "ce": ce.item(), "t0": t0_loss.item()}
+
+        self.log("train", metrics)
+
+        return loss
+
+    def validation_step(self, batch, batch_idx):
```
Comment (SpirinEgor) [inline src/diffusion/model.py:127]: What do you think of calculating accuracy on validation? Just to see how many tokens model predict correctly

File: src/diffusion/model.py [hunk 7]
```diff
@@ -64,3 +86,80 @@ def forward(
...
+            t == 1,
+            flat_mean((x_0_hat - emb) ** 2 * emb_mask),
+            flat_mean((x_0_hat - x_0) ** 2 * emb_mask),
+        ).mean()
+
+        t0_loss = (x_0**2 * emb_mask).mean()
+
+        loss = mse + ce + t0_loss
+
+        metrics = {"mse": mse.item(), "ce": ce.item(), "t0": t0_loss.item()}
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: Let's also log the total `loss`

**Review by SpirinEgor — CHANGES_REQUESTED** [review]
There are a lot of changes due to the black refactor. Please, use a line width of 120 to avoid this. For example, `dataset.py` should be without any changes in this PR.

File: src/diffusion/model.py [hunk 8]
```diff
@@ -64,3 +88,92 @@ def forward(
...
+        t0_loss = (x_0**2 * emb_mask).mean()
+
+        loss = mse + ce + t0_loss
+
+        metrics = {
+            "train/mse": mse.item(),
+            "train/ce": ce.item(),
+            "train/t0": t0_loss.item(),
+            "train/loss": loss.item(),
+        }
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: Lightning recommend to avoid `.item()` calls (https://pytorch-lightning.readthedocs.io/en/stable/guides/speed.html#things-to-avoid). Just pass tensors here:
```python
losses = {"train/mse": mse, "train/ce": ce, "train/t0": t0_loss, "train/loss": loss }
```

File: src/diffusion/model.py [hunk 9]
```diff
@@ -64,3 +88,92 @@ def forward(
...
+        pad_mask = gt == self.emb.padding_idx
+        probs = self.classifier(x_0)
+        ce = calculate_batch_ce(probs, gt, pad_mask)
+
+        emb_mask = ~pad_mask
+        preds = probs.argmax(-1)
+
+        acc = (((preds == gt) * emb_mask).sum(axis=1) / emb_mask.sum(axis=1)).mean()
+
+        return torch.tensor([ce.item(), acc.item()])
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: No need to create tensor here and call `.item()`. You can use `.detach()`, but I don't think that this would speed up. Return here just tuple: `return ce, acc`

File: src/training.py [hunk 10]
```diff
@@ -1,80 +1,48 @@
...
+        log_every_n_steps=logging_step,
+        logger=wandb_logger,
+        max_steps=num_steps,
+        val_check_interval=val_interval,
+        check_val_every_n_epoch=None,
+    )
+
+    trainer.fit(model, train_dataloaders=train_dataloader, val_dataloaders=val_dataloader)
+
+    wandb.finish()
```
Comment (SpirinEgor) [inline src/training.py]: Hmm, I'm not sure, but I think this is unnecessary

**Review by SpirinEgor — APPROVED** [review]
👍
